### PR TITLE
Support complex type in migrate procedure

### PR DESCRIPTION
**Pull request by chenjian2664**
## Description
Fix https://github.com/trinodb/trino/issues/17583

## Additional context and related issues
https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadata.java#L103 Will reassign columnId start from 1, we need to reset `nextFieldId`.

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Add support for `array`, `map` and `row` types in migrate procedure. ({issue}`17583`)
```

### CODE REVIEW

**Review by ebyhr — COMMENTED** [review]
Please add a product test.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 1]
```diff
@@ -59,19 +60,32 @@ public void testMigrateTable(IcebergFileFormat fileFormat)
...
+        assertUpdate("CREATE TABLE " + hiveTableName + " WITH (format='" + fileFormat + "') AS SELECT 1 x, array[2, 3] a, CAST(map(array['key1'], array['value1']) AS map(varchar, varchar)) b, CAST(row(1) AS row(d integer)) c", 1);
         assertQueryFails("SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
 
         assertUpdate("CALL iceberg.system.migrate('tpch', '" + tableName + "')");
 
         assertThat((String) computeScalar("SHOW CREATE TABLE " + icebergTableName))
                 .contains("format = '%s'".formatted(fileFormat));
 
-        assertQuery("SELECT * FROM " + icebergTableName, "VALUES 1");
+        @Language("SQL") String firstRow = "VALUES (" +
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:71]: Please revert and add a different test method.

**Comment by chenjian2664** [comment]
> Please add a product test.

Could you suggest how to represent the `array`, `map`, `row` value in `TestIcebergProcedureCalls`, I am confusing about this part

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 2]
```diff
@@ -56,6 +63,34 @@ public void testMigrateHiveTable()
         onTrino().executeQuery("DROP TABLE IF EXISTS " + icebergTableName);
     }
 
+    @Test(groups = {ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testMigrateHiveTableWithComplexType()
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:67]: Thanks for adding the product test. Please test with all file format (orc, parquet, avro) because Spark behavior might be different based on that.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java [hunk 3]
```diff
@@ -297,24 +293,17 @@ private Schema toIcebergSchema(List<Column> columns)
         List<Types.NestedField> icebergColumns = new ArrayList<>();
         for (Column column : columns) {
             int index = icebergColumns.size();
-            org.apache.iceberg.types.Type type = toIcebergType(typeManager.getType(column.getType().getTypeSignature()), nextFieldId);
+            org.apache.iceberg.types.Type type = toIcebergTypeForNewColumn(typeManager.getType(column.getType().getTypeSignature()), nextFieldId);
             Types.NestedField field = Types.NestedField.of(index, false, column.getName(), type, column.getComment().orElse(null));
             icebergColumns.add(field);
         }
+
+        nextFieldId.set(1);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java:301]: This "reset" operation is rather hard to follow.

Ideally we shouldn't have the following call of `TypeUtil.assignFreshIds(...)` because we've already done the assignment previously.
At the moment, the IDs which are assigned in the loop are fully discarded. At the very least, I'd recommend using for a better readability (even a comment would be welcomed):

```
icebergSchema = TypeUtil.assignFreshIds(icebergSchema, new AtomicInteger(1)::getAndIncrement);
```


I've did 

```
    private Schema toIcebergSchema(List<Column> columns)
    {
        AtomicInteger nextFieldId = new AtomicInteger(1);
        List<Types.NestedField> icebergColumns = new ArrayList<>();
        for (Column column : columns) {
            Types.NestedField field = Types.NestedField.of(nextFieldId.getAndIncrement(), false, column.getName(), toIcebergTypeForNewColumn(typeManager.getType(column.getType().getTypeSignature()), nextFieldId), column.getComment().orElse(null));
            icebergColumns.add(field);
        }

        org.apache.iceberg.types.Type icebergSchema = Types.StructType.of(icebergColumns);
        return new Schema(icebergSchema.asStructType().fields());
    }
```

However the tests fail, because, even though the fields get unique ids, `TypeUtil.assignFreshIds` works in a more breadth first fashion while the current logic of assigning IDs (via `toIcebergTypeForNewColumn` is depth-first.

Comment (chenjian2664) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java]: Thanks for the point, I have thought about using breadth first way to build the columns, but feels a little bit overkill for this method and the readability may not even better, WDYT?

File: testing/trino-product-tests/pom.xml [hunk 4]
```diff
@@ -150,6 +150,16 @@
             <artifactId>java-driver-core</artifactId>
         </dependency>
 
+        <dependency>
```
Comment (findinpath) [inline testing/trino-product-tests/pom.xml:153]: Why do we have the new dependencies as part of this PR?

Comment (chenjian2664) [reply testing/trino-product-tests/pom.xml:153]: Product test `TestIcebergProcedureCalls.testMigrateHiveTableWithComplexType` we use the dependency to parse spark result.

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/actions/runs/5319631815/jobs/9632630276?pr=17779#step:7:62380 Is the sth wrong with delta lake tests?

**Comment by findinpath** [comment]
@chenjian2664  pls rebase on top of `master` to address the code conflicts.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 5]
```diff
@@ -76,6 +77,43 @@ public void testMigrateTable(IcebergFileFormat fileFormat)
...
+                "CAST(row(2) AS row(d integer)))";
+        assertUpdate("INSERT INTO " + icebergTableName + secondRow, 1);
+        assertQuery("SELECT count(*) FROM " + icebergTableName, "VALUES 2");
+        assertThat(query("SELECT * FROM " + icebergTableName))
+                .matches(firstRow + " UNION ALL " + secondRow);
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
     @DataProvider
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:117]: Consider adding the following test cases to your PR

```
    @Test
    public void testMigrateTableParquetSchemaEvolution()
            throws Exception
    {
        String randomNameSuffix = randomNameSuffix();
        String tableNameOneColumn = "test_migrate_one_column_" + randomNameSuffix;
        String tableNameTwoColumns = "test_migrate_two_columns_" + randomNameSuffix;
        String hiveTableNameOneColumn = "hive.tpch." + tableNameOneColumn;
        String hiveTableNameTwoColumns = "hive.tpch." + tableNameTwoColumns;
        String icebergTableNameTwoColumns = "iceberg.tpch." + tableNameTwoColumns;

        assertUpdate("CREATE TABLE " + hiveTableNameOneColumn + " WITH (format = 'PARQUET') AS SELECT 1 col1", 1);
        assertUpdate("CREATE TABLE " + hiveTableNameTwoColumns + " WITH (format = 'PARQUET') AS SELECT 2 col1, CAST(row(10, 20) AS row(x integer, y integer)) AS nested", 1);

        // Copy the parquet file containing only one column to the table with two columns
        Path tableNameOneColumnLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameOneColumn));
        Path tableNameTwoColumnsLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameTwoColumns));
        try (Stream<Path> files = Files.list(tableNameOneColumnLocation)) {
            Path file = files.filter(path -> !path.getFileName().toString().startsWith(".")).collect(onlyElement());
            Files.copy(file, tableNameTwoColumnsLocation.resolve(file.getFileName()));
        }

        assertUpdate("CALL iceberg.system.migrate('tpch', '" + tableNameTwoColumns + "')");

        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20))");

        assertUpdate("INSERT INTO " + icebergTableNameTwoColumns + " VALUES (3, row(100, 200))", 1);
        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20)), (3, row(100, 200))");

        assertUpdate("DROP TABLE " + icebergTableNameTwoColumns);
    }

    @Test
    public void testMigrateTableParquetRowColumnSchemaEvolution()
            throws Exception
    {
        String randomNameSuffix = randomNameSuffix();
        String tableNameRowOneColumn = "test_migrate_row_one_column_" + randomNameSuffix;
        String tableNameRowTwoColumns = "test_migrate_row_two_columns_" + randomNameSuffix;
        String hiveTableNameRowOneColumn = "hive.tpch." + tableNameRowOneColumn;
        String hiveTableNameRowTwoColumns = "hive.tpch." + tableNameRowTwoColumns;
        String icebergTableNameRowTwoColumns = "iceberg.tpch." + tableNameRowTwoColumns;

        assertUpdate("CREATE TABLE " + hiveTableNameRowOneColumn + " WITH (format = 'PARQUET') AS SELECT CAST(row(1) AS row(x integer)) as nested", 1);
        assertUpdate("CREATE TABLE " + hiveTableNameRowTwoColumns + " WITH (format = 'PARQUET') AS SELECT CAST(row(10, 20) AS row(x integer, y integer)) AS nested", 1);

        Path tableNameRowOneColumnLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameRowOneColumn));
        Path tableNameRowTwoColumnsLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameRowTwoColumns));
        try (Stream<Path> files = Files.list(tableNameRowOneColumnLocation)) {
            Path file = files.filter(path -> !path.getFileName().toString().startsWith(".")).collect(onlyElement());
            Files.copy(file, tableNameRowTwoColumnsLocation.resolve(file.getFileName()));
        }

        assertUpdate("CALL iceberg.system.migrate('tpch', '" + tableNameRowTwoColumns + "')");

        assertThat(query("SELECT * FROM " + icebergTableNameRowTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES row(CAST((1,null) AS row(x integer, y integer))), row(row(10, 20))");

        assertUpdate("INSERT INTO " + icebergTableNameRowTwoColumns + " VALUES (row(row(100, 200)))", 1);
        assertThat(query("SELECT * FROM " + icebergTableNameRowTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES row(CAST((1, null) AS row(x integer, y integer))), row(row(10, 20)), row(row(100, 200))");

        assertUpdate("DROP TABLE " + icebergTableNameRowTwoColumns);
    }

```

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java (hunk 5 above)
Comment (chenjian2664) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:188]: handsome

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 6]
```diff
@@ -76,6 +77,114 @@ public void testMigrateTable(IcebergFileFormat fileFormat)
...
+
+        assertUpdate("DROP TABLE " + icebergTableNameTwoColumns);
+    }
+
+    @Test
+    public void testMigrateTableParquetRowColumnSchemaEvolution()
+            throws Exception
+    {
+        String randomNameSuffix = randomNameSuffix();
+        String tableNameRowOneColumn = "test_migrate_row_one_column_" + randomNameSuffix;
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:158]: Sorry, I've made a little mistake yesterday

`Column` should be renamed to `Field`

https://trino.io/docs/current/language/types.html#row

The row has fields, not columns.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 7]
```diff
@@ -76,6 +77,114 @@ public void testMigrateTable(IcebergFileFormat fileFormat)
...
+        assertUpdate("INSERT INTO " + icebergTableNameTwoColumns + " VALUES (3, row(100, 200))", 1);
+        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
+                .skippingTypesCheck()
+                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20)), (3, row(100, 200))");
+
+        assertUpdate("DROP TABLE " + icebergTableNameTwoColumns);
+    }
+
+    @Test
+    public void testMigrateTableParquetRowColumnSchemaEvolution()
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:154]:
```suggestion
    @Test(dataProvider = "fileFormats")
    public void testMigrateTableRowColumnTypeEvolution(IcebergFileFormat fileFormat)
```

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 8]
```diff
@@ -76,6 +77,114 @@ public void testMigrateTable(IcebergFileFormat fileFormat)
...
+        assertUpdate("INSERT INTO " + icebergTableName + secondRow, 1);
+        assertQuery("SELECT count(*) FROM " + icebergTableName, "VALUES 2");
+        assertThat(query("SELECT * FROM " + icebergTableName))
+                .matches(firstRow + " UNION ALL " + secondRow);
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
+    @Test
+    public void testMigrateTableParquetSchemaEvolution()
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:118]:
```suggestion
    @Test(dataProvider = "fileFormats")
    public void testMigrateTableSchemaEvolution(IcebergFileFormat fileFormat)
```

**Comment by chenjian2664** [comment]
It's strange that my commits is ignored by git..

**Comment by findinpath** [comment]
@chenjian2664  pls push an empty commit.

@ebyhr / @findepi  can you please do a check on this PR?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java [hunk 9]
```diff
@@ -338,8 +339,8 @@ private Expression unwrapCast(ComparisonExpression expression)
                             yield new ComparisonExpression(LESS_THAN_OR_EQUAL, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType));
                         }
                         case GREATER_THAN, GREATER_THAN_OR_EQUAL ->
-                                // We expect implicit coercions to be order-preserving, so the result of converting back from target -> source cannot produce a value
-                                // larger than the next value in the source type
+                            // We expect implicit coercions to be order-preserving, so the result of converting back from target -> source cannot produce a value
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:342]: @chenjian2664  why do you have now the extra commits?

Comment (chenjian2664) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:342]: Strange ...
Let me check

**Comment by findepi** [comment]
@marcinsbd PTAL

**Comment by chenjian2664** [comment]
@findinpath @ebyhr @findepi Could you please take a look?

**Comment by chenjian2664** [comment]
Does anyone can help on this? @findepi

**Comment by mosabua** [comment]
@ebyhr @findepi .. could you chime in here and potentially merge?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 10]
```diff
@@ -104,6 +104,43 @@ public void testMigrateTableWithTinyintType(IcebergFileFormat fileFormat)
...
+
+    @ParameterizedTest
+    @MethodSource("fileFormats")
+    public void testMigrateTableWithComplexType(IcebergFileFormat fileFormat)
+    {
+        String tableName = "test_migrate_complex_" + randomNameSuffix();
+        String hiveTableName = "hive.tpch." + tableName;
+        String icebergTableName = "iceberg.tpch." + tableName;
+
+        assertUpdate("CREATE TABLE " + hiveTableName + " WITH (format='" + fileFormat + "') AS SELECT 1 x, array[2, 3] a, CAST(map(array['key1'], array['value1']) AS map(varchar, varchar)) b, CAST(row(1) AS row(d integer)) c", 1);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:118]: This line is too long.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 11]
```diff
@@ -136,6 +173,79 @@ public void testMigrateTableWithSmallintType(IcebergFileFormat fileFormat)
...
+        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
+                .skippingTypesCheck()
+                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20))");
+
+        assertUpdate("INSERT INTO " + icebergTableNameTwoColumns + " VALUES (3, row(100, 200))", 1);
+        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
+                .skippingTypesCheck()
+                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20)), (3, row(100, 200))");
+
+        assertUpdate("DROP TABLE " + icebergTableNameTwoColumns);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:210]: Drop hiveTableNameOneColumn table.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 12]
```diff
@@ -136,6 +173,79 @@ public void testMigrateTableWithSmallintType(IcebergFileFormat fileFormat)
...
+        assertThat(query("SELECT * FROM " + icebergTableNameRowTwoFields))
+                .skippingTypesCheck()
+                .matches("VALUES row(CAST((1,null) AS row(x integer, y integer))), row(row(10, 20))");
+
+        assertUpdate("INSERT INTO " + icebergTableNameRowTwoFields + " VALUES (row(row(100, 200)))", 1);
+        assertThat(query("SELECT * FROM " + icebergTableNameRowTwoFields))
+                .skippingTypesCheck()
+                .matches("VALUES row(CAST((1, null) AS row(x integer, y integer))), row(row(10, 20)), row(row(100, 200))");
+
+        assertUpdate("DROP TABLE " + icebergTableNameRowTwoFields);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:246]: Drop hiveTableNameRowOneField table.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 13]
```diff
@@ -111,6 +119,38 @@ public void testMigrateHiveTableWithSmallintType(String fileFormat)
...
+        onTrino().executeQuery("CALL iceberg.system.migrate('default', '" + tableName + "')");
+
+        List<Object> expectedRowValue = ImmutableList.of(1, ImmutableList.of(2, 3), ImmutableMap.of("key", "value"), ImmutableMap.of("d", 1));
+
+        List<JDBCType> trinoExpectedColumnTypes = ImmutableList.of(INTEGER, ARRAY, JAVA_OBJECT, JAVA_OBJECT);
+        String trinoExpectedRowValueString = expectedRowValue.toString();
+        assertThat(onTrino().executeQuery("SELECT * FROM " + icebergTableName))
+                .hasRowsCount(1)
+                .matches(queryResult -> queryResult.getColumnTypes().equals(trinoExpectedColumnTypes), "Complex type columns type mismatch in Trino")
+                .matches(queryResult -> queryResult.row(0).toString().equals(trinoExpectedRowValueString), "Complex type columns value mismatch in Trino");
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:142]: Please verify each values instead of string. Same for Spark assertion.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 14]
```diff
@@ -136,6 +173,79 @@ public void testMigrateTableWithSmallintType(IcebergFileFormat fileFormat)
...
+        String tableNameOneColumn = "test_migrate_one_column_" + randomNameSuffix;
+        String tableNameTwoColumns = "test_migrate_two_columns_" + randomNameSuffix;
+        String hiveTableNameOneColumn = "hive.tpch." + tableNameOneColumn;
+        String hiveTableNameTwoColumns = "hive.tpch." + tableNameTwoColumns;
+        String icebergTableNameTwoColumns = "iceberg.tpch." + tableNameTwoColumns;
+
+        assertUpdate("CREATE TABLE " + hiveTableNameOneColumn + " WITH (format='" + fileFormat + "') AS SELECT 1 col1", 1);
+        assertUpdate("CREATE TABLE " + hiveTableNameTwoColumns + " WITH (format='" + fileFormat + "') AS SELECT 2 col1, CAST(row(10, 20) AS row(x integer, y integer)) AS nested", 1);
+
+        // Copy the parquet file containing only one column to the table with two columns
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:191]: This test isn't specific to Parquet. 
```suggestion
        // Copy the file containing only one column to the table with two columns
```

**Review by ebyhr — COMMENTED** [review]
Looks good except for comments.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java [hunk 15]
```diff
@@ -104,7 +104,6 @@ public void testMigrateTableWithTinyintType(IcebergFileFormat fileFormat)
 
         assertUpdate("INSERT INTO " + icebergTableName + " VALUES -2147483648, 2147483647", 2);
         assertQuery("SELECT * FROM " + icebergTableName, "VALUES (NULL), (-2147483648), (-128), (127), (2147483647)");
-
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java:107]: Please revert unrelated change.

**Comment by chenjian2664** [comment]
2024-01-26 17:02:36 INFO: FAILURE     /    io.trino.tests.product.iceberg.TestIcebergProcedureCalls.testMigrateHiveTableWithComplexType [AVRO] (Groups: profile_specific_tests, iceberg) took 0.5 seconds
tests               | 2024-01-26 17:02:36 SEVERE: Failure cause:
tests               | java.lang.AssertionError: Could not find rows:
tests               | [1, [2, 3], {key=value}, {d=1}]
tests               | 
tests               | actual rows:
tests               | [1, [2, 3], {key=value}, {d=1}]
tests               | 	at io.trino.tests.product.iceberg.TestIcebergProcedureCalls.testMigrateHiveTableWithComplexType(TestIcebergProcedureCalls.java:133)
tests               | 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:1583)
tests               | 
tests               | 2024-01-26 17:02:36 INFO: [9 of 101] io.trino.tests.product.iceberg.TestIcebergProcedureCalls.testMigrateHiveTableWithSmallintType [ORC] (Groups: profile_specific_tests, iceberg)

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java (hunk 13 above)
Comment (chenjian2664) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java]: Updated the implementation. Please suggests if you have any better ways, thanks!

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 16]
```diff
@@ -111,6 +115,46 @@ public void testMigrateHiveTableWithSmallintType(String fileFormat)
...
+
+        assertThat(onTrino().executeQuery("SELECT * FROM " + icebergTableName))
+                .satisfies(queryResult -> {
+                    assertThat(queryResult).hasRowsCount(1).hasColumnsCount(4);
+                    List<?> row = queryResult.row(0);
+
+                    assertThat(row.get(0)).isEqualTo(1);
+                    assertThat(row.get(1)).isInstanceOf(TrinoArray.class);
+                    assertThat(((TrinoArray) row.get(1)).getArray()).isEqualTo(new Object[] {2, 3});
+                    assertThat(row.get(2)).isEqualTo(ImmutableMap.of("key", "value"));
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:141]: We could simply retrieve elements with subscript operators like `SELECT x, a[0], ...`. That's how we verify complex types in other existing product tests.

**Review by ebyhr — APPROVED** [review]
